src/Workspaces/Core/Portable/Workspace/WorkspaceRegistration.cs

```diff
@@ -23,12 +24,23 @@ internal void SetWorkspaceAndRaiseEvents(Workspace workspace)
 
         internal void SetWorkspace(Workspace workspace)
         {
-            _registeredWorkspace = workspace;
+            Workspace = workspace;
         }
 
         internal void RaiseEvents()
         {
-            WorkspaceChanged?.Invoke(this, EventArgs.Empty);
+            lock (_gate)
```

Comment:
RaiseEvents is also called externally to this type -- note it's not private! (and yes, ugh.)

Reply:
why put gate there? issue is on event not on setting workspace.

Reply:
💭 It seems like another option is to update `UnregisterText` to wait for the queue to complete before calling `SetWorkspaceAndRaiseEvents`.

Reply:
> why put gate there? issue is on event not on setting workspace.

I believe the current form allows the following sequence:

1. `RegisterText` sets the workspace to non-null
2. `UnregisterText` sets the workspace to null
3. `WorkspaceChanged` is invoked, and the workspace is null
4. `WorkspaceChanged` is invoked again, and the workspace is still null

Reply:
Option 2 is something we should consider, however it's somewhat risky for Preview 2 given we don't know who is expecting the event to be raised synchronously during a Workspace unhook from a buffer. We [serviced Dev14 for a bug like that](https://github.com/dotnet/roslyn/pull/20319), hence I'd advise some caution. I do agree that it's odd that RegisterText and UnregisterText use different approaches; we ran into this years ago and I think concluded it was just an oops when one got updated and the other wasn't. I'm entirely for making them match...in Preview 3 or 16.1. :smile:

Keep in mind that different workspaces (with different event queues!) can be potentially interacting with this type at the same time. One workspace might be unregistering itself while another workspace is registering itself -- this potentially happens in the case of somebody taking ownership of a document in the RDT, and the MiscellaneousFilesWorkspace then wants to unhook itself. That makes queue management tricky.

Option 3 is intriguing.

Also looking at this code a bit more, there's another bug: the code that calls UnregisterText verifies it'll only call UnregisterText if currently associated with itself, so you don't accidentally unregister somebody else at the same time. But it really should be doing what is effectively a compare-and-set under a lock, so @sharwell's point about expanding the lock does stand; it should also be in SetWorkspaceAndRaiseEvents, and a SetWorkspaceIfCurrentWorkspaceAndRaiseEvents() would also exist in replacement to the SetWorkspace/Raise method that exists today.

Reply:
I dont see anything else than making event not raised concurrently is meaningful unless we fix whole issues all together listed in the linked issue.

right now, since event is raised one in async fashion and the other in sync fashion, but workspace is always set in sync fashion already make workspace not something event listener can reason about.

if we want to make workspace and event to be tightly coupled, workspace should have given as argument of the event such as old Workpsace and new Workspace like how WorkspaceEvent does for solution. otherwise, workspace property still not that useful for the event listener.

anyway, I dont want to change whole that design in this PR. I opened new issue to fix this properly. I dont believe we should just leave things as they are just because we didn't come up with perfect solution yet.

if anyone has strong opinion or perfect solution for the issue, please put that one in the linked issue. and set priority of the issue following one's belief how urgent, important the issue is.